File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.swift
import Foundation

func solution(_ cap:Int, _ n:Int, _ deliveries:[Int], _ pickups:[Int]) -> Int64 {
    
    var d = 0
    var p = 0
    var move = 0
    
    for i in stride(from: n-1, through: 0, by: -1) {
        
        d += deliveries[i]
        p += pickups[i]
        
        while d > 0 || p > 0 {
            
            d -= cap
            p -= cap
            
            move += (i+1) * 2
        }
    }
    
    return Int64(move)
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
# @time 25

def deliver(deliveries, k):
        
    while deliveries and k > 0:
        if deliveries[-1] > 0:
            if deliveries[-1] > k:
                deliveries[-1] -= k
                k = 0
            else:
                k -= deliveries[-1]
                deliveries.pop()
        else:
            deliveries.pop()
            
    while deliveries and deliveries[-1] == 0:
        deliveries.pop()
            
def pickup(pickups, k):
    
    while pickups and k > 0:
        if pickups[-1] > 0:
            if pickups[-1] > k:
                pickups[-1] -= k
                k = 0
            else:
                k -= pickups[-1]
                pickups.pop()
        else:
            pickups.pop()
            
    while pickups and pickups[-1] == 0:
        pickups.pop()

def solution(cap, n, deliveries, pickups):
    
    answer = 0
    
    while deliveries and deliveries[-1] == 0:
        deliveries.pop()
    
    while pickups and pickups[-1] == 0:
        pickups.pop()
    
    answer = max(len(deliveries), len(pickups)) * 2
    
    while deliveries or pickups:
        deliver(deliveries, cap)
        pickup(pickups, cap)
        
        answer += max(len(deliveries), len(pickups)) * 2
        
        
        
    return answer